grudge/flux_differencing.py

```diff
@@ -0,0 +1,264 @@
+    if not isinstance(flux_matrix, DOFArray):
+        return map_array_container(
+            partial(_single_axis_hybridized_sbp_derivative_kernel,
+                    dcoll, dq, df, xyz_axis),
+            flux_matrix
+        )
+
+    from grudge.geometry import \
+        area_element, inverse_surface_metric_derivative
+    from grudge.interpolation import (
+        volume_and_surface_interpolation_matrix,
+        volume_and_surface_quadrature_interpolation
+    )
+
+    # FIXME: This is kinda meh
+    def inverse_jac_matrix():
+        @memoize_in(dcoll, (_single_axis_hybridized_sbp_derivative_kernel, dq, df))
+        def _inv_surf_metric_deriv():
+            mat = actx.np.stack(
+                [
+                    actx.np.stack(
+                        [
+                            volume_and_surface_quadrature_interpolation(
+                                dcoll, dq, df,
+                                area_element(actx, dcoll)
+                                * inverse_surface_metric_derivative(
+                                    actx, dcoll,
+                                    rst_ax, xyz_ax
+                                )
+                            ) for rst_ax in range(dcoll.dim)
+                        ]
+                    ) for xyz_ax in range(dcoll.ambient_dim)
+                ]
+            )
+            return freeze(mat, actx)
+        return _inv_surf_metric_deriv()
+
+    actx = flux_matrix.array_context
+    discr = dcoll.discr_from_dd("vol")
+    quad_volm_discr = dcoll.discr_from_dd(dq)
+    quad_face_discr = dcoll.discr_from_dd(df)
+
+    # FIXME: This assumes affine-geometry
+    return DOFArray(
+        actx,
+        data=tuple(
+            # r for rst axis
+            actx.einsum("ik,rej,rij,eij->ek",
+                        volume_and_surface_interpolation_matrix(
+                            actx,
+                            base_element_group=bgrp,
+                            vol_quad_element_group=qvgrp,
+                            face_quad_element_group=qafgrp,
+                            dtype=fmat_i.dtype
+                        ),
+                        ijm_i[xyz_axis],
+                        skew_symmetric_hybridized_sbp_operators(
+                            actx,
+                            bgrp,
+                            qvgrp,
+                            qafgrp,
+                            fmat_i.dtype
+                        ),
+                        fmat_i,
+                        arg_names=("Vh_mat_t", "inv_jac_t", "ref_Q_mat", "F_mat"),
+                        tagged=(FirstAxisIsElementsTag(),))
+
+            for bgrp, qvgrp, qafgrp, fmat_i, ijm_i in zip(
+                discr.groups,
+                quad_volm_discr.groups,
+                quad_face_discr.groups,
+                flux_matrix,
+                inverse_jac_matrix()
+            )
+        )
+    )
+
+
+def _flux_differencing_helper(dcoll, diff_func, mats):
```

Comment:
Possible to reuse `div_helper`? (If not, is there an extension of `div_helper` that might do the job?)

Reply:
@thomasgibson says it's the same thing.

Reply:
Done: https://github.com/inducer/grudge/pull/214/commits/9ba57df378663ca65504d839789a017bf26a0d55